libraries/ms-common/src/can.c

```diff
@@ -0,0 +1,155 @@
+#include "can.h"
+#include "soft_timer.h"
+#include "can_rx_fsm.h"
+
+#define CAN_BUS_OFF_RECOVERY_TIME_MS 500
+
+// Attempts to transmit the specified message using the HW TX, overwriting the source device.
+StatusCode prv_transmit(const CANConfig *can, const CANMessage *msg) {
+  CANId msg_id = {
+    .source_id = can->device_id,
+    .type = msg->type,
+    .msg_id = msg->msg_id
+  };
+
+  return can_hw_transmit(&can->hw, msg_id.raw, msg->data_u8, msg->dlc);
+}
+
+// Handler for CAN HW TX ready events
+// Attempts to transmit any messages in the TX queue
+void prv_tx_handler(void *context) {
+  CANConfig *can = context;
+  CANMessage tx_msg;
+
+  while (can_queue_size(&can->tx_queue) > 0) {
+    can_queue_peek(&can->tx_queue, &tx_msg);
+
+    StatusCode result = prv_transmit(can, &tx_msg);
+
+    if (result != STATUS_CODE_OK) {
```

Comment:
`can_hw_transmit` returns a statuscode if transmit fails, so that should be enough. A status callback should be good for alerting us about what's gone wrong.